source/server/connection_handler_impl.cc

```diff
@@ -54,6 +68,31 @@ void ConnectionHandlerImpl::removeListeners(uint64_t listener_tag) {
   }
 }
 
+void ConnectionHandlerImpl::removeFilterChains(
+    Network::DrainingFilterChains& draining_filter_chains, std::function<void()> completion) {
+  for (auto& listener : listeners_) {
+    // TODO(lambdai):
+    if (listener.second.listener_->listenerTag() ==
+        draining_filter_chains.getDrainingListenerTag()) {
+      ASSERT(!listener.second.tcp_listener_.has_value());
+      listener.second.tcp_listener_->get().removeFilterChains(
+          draining_filter_chains.getDrainingFilterChains());
+      Event::DeferredTaskUtil::deferredRun(dispatcher_, std::move(completion));
+      return;
+    }
+  }
+  // Fallback to iterate over all listeners. The reason is that the target listener might have began
+  // another update and the previous tag is lost.
+  // TODO(lambdai): lookup the correct listener
```

Comment:
?

Reply:
This is introduced in this PR: A worker thread active listener is somewhat RCU style, instead of 
1. The listener are updates from 1st config to 2nd and to 3rd, now worker thread sees the 3rd listener tag
2. At this comment, the 2nd listener drain timeout is triggered and now worker need to drain filter chains of **2nd** listener.

Ideally the production code is to track the tags of listener, where the draining listener tag was 2 but connection handler should drain listener filter chains from listener tag 3